File: watch/wearos/src/main/java/app/vescape/wear/WatchWeather.kt
package app.vescape.wear

import androidx.annotation.DrawableRes
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableLongStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import kotlinx.coroutines.delay
import java.util.Calendar

/**
 * Data Layer path the phone publishes the forecast on. Must match the phone-side
 * `WatchWeatherPusher`. Arrives at most once per forecast refresh and then persists, so it is read
 * on every start rather than waited for.
 *
 * @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/watch/WatchWeather.kt
 */
const val WEATHER_PATH = "/weather"

const val WEATHER_TEMP_C = "temperatureC"
const val WEATHER_ICON = "icon"
const val WEATHER_LABEL = "label"
const val WEATHER_PRECIP = "precipitationProbability"
const val WEATHER_HOUR_MINUTES = "hourMinutes"
const val WEATHER_HOUR_TEMPS = "hourTemps"
const val WEATHER_HOUR_ICONS = "hourIcons"
const val WEATHER_HOUR_PRECIPS = "hourPrecips"
const val WEATHER_SUNRISE = "sunriseMinuteOfDay"
const val WEATHER_SUNSET = "sunsetMinuteOfDay"
const val WEATHER_LATITUDE = "latitude"
const val WEATHER_LONGITUDE = "longitude"
const val WEATHER_FETCHED_AT = "fetchedAtMs"

/** One forecast hour, as the wrist renders it. [minuteOfDay] is local to the forecast location. */
data class WatchWeatherHour(
    val minuteOfDay: Int,
    val temperatureC: Int,
    val icon: String,
    val precipitationProbability: Int,
)

/**
 * The forecast on the wrist. Absent until the phone has pushed one — the wrist never fetches, and a
 * phone that has not seen a GPS Fix yet has nothing to send.
 */
data class WatchWeather(
    val temperatureC: Int,
    val icon: String,
    val label: String,
    val precipitationProbability: Int,
    val hourly: List<WatchWeatherHour>,
    /** Minutes since local midnight; null when the phone had no daily times to send. */
    val sunriseMinuteOfDay: Int?,
    val sunsetMinuteOfDay: Int?,
    /**
     * Where the forecast was taken. Null from a phone too old to send it, which is why the radar
     * page can be empty on a wrist that is otherwise showing weather fine.
     */
    val latitude: Double?,
    val longitude: Double?,
    val fetchedAtMs: Long,
)

/** Latest forecast pushed from the phone; null until the first push arrives. */
object WeatherState {
    val weather = mutableStateOf<WatchWeather?>(null)

    fun accept(weather: WatchWeather?) {
        this.weather.value = weather
    }
}

/**
 * How long a pushed forecast is worth showing. The Data Layer keeps the last item forever, so a
 * phone that stopped refreshing (app killed, out of range, GPS off) would otherwise leave yesterday's
 * conditions on the wrist looking current. Generous next to the phone's ten-minute refresh: this is
 * the line between "a bit old" and "not weather any more".
 */
const val WEATHER_STALE_MS = 3 * 60 * 60 * 1_000L

/**
 * The pushed forecast while it is still worth believing, else null. Recomposes on its own clock so
 * an aged-out reading disappears without waiting for a push that is never coming.
 */
@Composable
fun freshWeather(): WatchWeather? {
    val weather by WeatherState.weather
    var nowMs by remember { mutableLongStateOf(System.currentTimeMillis()) }

    LaunchedEffect(Unit) {
        while (true) {
            delay(STALENESS_TICK_MS)
            nowMs = System.currentTimeMillis()
        }
    }

    val forecast = weather ?: return null
    // A clock that jumped backwards (timezone/NTP correction) must not read as "from the future" and
    // hide a forecast that is fine; only real age hides it.
    return if (nowMs - forecast.fetchedAtMs > WEATHER_STALE_MS) null else forecast
}

/** Coarse: the readout only has to drop within a minute or so of the forecast ageing out. */
private const val STALENESS_TICK_MS = 60_000L

/**
 * Condition slug into the ported Phosphor drawable. The phone resolves the slug from the WMO code,
 * so the wrist never classifies weather — an unknown slug from a newer phone draws a plain cloud
 * rather than nothing.
 *
 * @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/weather/Weather.kt `WeatherIcon`
 */
@DrawableRes
fun weatherIconRes(slug: String): Int = when (slug) {
    "sun" -> R.drawable.ic_ph_sun
    "moon" -> R.drawable.ic_ph_moon_stars
    "cloud-sun" -> R.drawable.ic_ph_cloud_sun
    "cloud-moon" -> R.drawable.ic_ph_cloud_moon
    "cloud-fog" -> R.drawable.ic_ph_cloud_fog
    "cloud-rain" -> R.drawable.ic_ph_cloud_rain
    "cloud-snow" -> R.drawable.ic_ph_cloud_snow
    "cloud-lightning" -> R.drawable.ic_ph_cloud_lightning
    else -> R.drawable.ic_ph_cloud
}

/**
 * Local wall-clock minute of day, re-read on the same coarse tick as staleness so a forecast left
 * open on the wrist crosses sunset on its own.
 */
@Composable
fun currentMinuteOfDay(): Int {
    var minuteOfDay by remember { mutableIntStateOf(minuteOfDayNow()) }

    LaunchedEffect(Unit) {
        while (true) {
            delay(STALENESS_TICK_MS)
            minuteOfDay = minuteOfDayNow()
        }
    }

    return minuteOfDay
}

/** Minute of day from the device clock, the unit every sun and hour time on the wrist is in. */
fun minuteOfDayNow(): Int = Calendar.getInstance().let {
    it.get(Calendar.HOUR_OF_DAY) * 60 + it.get(Calendar.MINUTE)
}

/** Local minute of day for an instant, the unit every wrist time is formatted from. */
fun minuteOfDay(epochMs: Long): Int = Calendar.getInstance().let {
    it.timeInMillis = epochMs
    it.get(Calendar.HOUR_OF_DAY) * 60 + it.get(Calendar.MINUTE)
}

/** `HH:MM` for a minute-of-day, matching [WatchClock]'s always-24h readout. */
fun formatHour(minuteOfDay: Int): String =
    String.format("%02d:%02d", (minuteOfDay / 60) % 24, minuteOfDay % 60)

File: modules/vescape-core/ios/watch/WatchWeather.swift
import Foundation

/// Channel the forecast occupies inside the shared Application Context (see `WatchColdState`).
///
/// Cold state like the settings: the forecast changes once per refresh at most — never per tick —
/// so it rides the Application Context and survives a restart or a reconnect instead of being
/// dropped like an undelivered message. Android publishes the same bag on its own Data Layer path.
///
/// @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/watch/WatchWeather.kt `WATCH_WEATHER_PATH`
/// @parity /watch/wearos/src/main/java/app/vescape/wear/WatchWeather.kt `WEATHER_PATH`
/// @platform-diff Android gets one Data Layer path per channel; watchOS has a single Application
///   Context, so `weather` is a key in the merged dictionary rather than a path of its own.
let watchWeatherChannel = "weather"

/// Wire keys, identical to Android's so the two wrists read the same bag.
///
/// @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/watch/WatchWeather.kt
/// @parity /watch/wearos/src/main/java/app/vescape/wear/WatchWeather.kt
enum WatchWeatherKey {
  /// Current temperature, whole degrees Celsius.
  static let temperatureC = "temperatureC"
  /// Condition pictogram slug, resolved by native so the wrist never classifies a WMO code.
  static let icon = "icon"
  /// One-line condition label, already phrased for a wrist-width readout.
  static let label = "label"
  /// Chance of precipitation right now, percent.
  static let precipitationProbability = "precipitationProbability"
  /// Forecast hours, packed as parallel arrays so the bag stays flat key-value.
  static let hourMinutes = "hourMinutes"
  static let hourTemps = "hourTemps"
  static let hourIcons = "hourIcons"
  static let hourPrecips = "hourPrecips"
  /// Sunrise / sunset as minutes since local midnight, absent when the forecast carried neither
  /// (polar day, or a provider response without daily times). Both or nothing: a lone sunrise
  /// reads as a bug.
  static let sunrise = "sunriseMinuteOfDay"
  static let sunset = "sunsetMinuteOfDay"
  /// Where the forecast was taken, so the wrist can ask a radar provider for imagery around the
  /// rider. The forecast location, not a live fix: it moves once per forecast refresh, which is the
  /// accuracy a radar frame kilometres across is drawn at anyway.
  static let latitude = "latitude"
  static let longitude = "longitude"
  /// When the forecast was fetched, so the wrist can age out a reading the phone stopped refreshing.
  static let fetchedAtMs = "fetchedAtMs"
}

/// One forecast hour, as the wrist renders it. `minuteOfDay` is local to the forecast location.
///
/// @parity /watch/wearos/src/main/java/app/vescape/wear/WatchWeather.kt `WatchWeatherHour`
struct WatchWeatherHour: Equatable {
  var minuteOfDay: Int
  var temperatureC: Int
  var icon: String
  var precipitationProbability: Int
}

/// The forecast on the wrist. Absent until the phone has pushed one — the wrist never fetches a
/// forecast, and a phone that has not seen a GPS Fix yet has nothing to send.
///
/// One file compiled into both the phone target and the watch target, so the writer and the reader
/// cannot drift; `WatchSettings.swift` already uses that arrangement, and Android has to duplicate
/// the same bag by convention across two Gradle modules.
///
/// @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/watch/WatchWeather.kt `WatchWeather`
/// @parity /watch/wearos/src/main/java/app/vescape/wear/WatchWeather.kt `WatchWeather`
struct WatchWeather {
  var temperatureC: Int
  var icon: String
  var label: String
  var precipitationProbability: Int
  var hourly: [WatchWeatherHour]
  /// Minutes since local midnight; nil when the phone had no daily times to send.
  var sunriseMinuteOfDay: Int?
  var sunsetMinuteOfDay: Int?
  /// Where the forecast was taken. Nil from a phone too old to send it, which is why the radar page
  /// can be empty on a wrist that is otherwise showing weather fine.
  var latitude: Double?
  var longitude: Double?
  var fetchedAtMs: Int64

  /// The wire bag. Parallel arrays rather than a list of dictionaries, matching the Wear OS
  /// `DataMap`: both ends already agree on the key names, and a flat bag is what makes an unknown
  /// key ignorable by an older wrist.
  ///
  /// @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/watch/WatchWeatherPusher.kt `push`
  var payload: [String: Any] {
    var payload: [String: Any] = [
      WatchWeatherKey.temperatureC: temperatureC,
      WatchWeatherKey.icon: icon,
      WatchWeatherKey.label: label,
      WatchWeatherKey.precipitationProbability: precipitationProbability,
      WatchWeatherKey.hourMinutes: hourly.map(\.minuteOfDay),
      WatchWeatherKey.hourTemps: hourly.map(\.temperatureC),
      WatchWeatherKey.hourIcons: hourly.map(\.icon),
      WatchWeatherKey.hourPrecips: hourly.map(\.precipitationProbability),
      WatchWeatherKey.fetchedAtMs: fetchedAtMs,
    ]
    // Omitted rather than sent as a sentinel: the wrist reads absence directly.
    if let sunriseMinuteOfDay { payload[WatchWeatherKey.sunrise] = sunriseMinuteOfDay }
    if let sunsetMinuteOfDay { payload[WatchWeatherKey.sunset] = sunsetMinuteOfDay }
    if let latitude { payload[WatchWeatherKey.latitude] = latitude }
    if let longitude { payload[WatchWeatherKey.longitude] = longitude }
    return payload
  }

  /// Read the bag leniently. A forecast without current conditions is nothing at all rather than a
  /// fabricated 0 °C clear sky — the same rule the phone's own parser follows for a short hourly
  /// array. The hours are parallel arrays, so an hour is kept only where every lane has a value.
  ///
  /// @parity /watch/wearos/src/main/java/app/vescape/wear/MainActivity.kt `readWeather`
  static func decode(_ payload: [String: Any]?) -> WatchWeather? {
    guard let payload,
      let temperatureC = (payload[WatchWeatherKey.temperatureC] as? NSNumber)?.intValue,
      let icon = payload[WatchWeatherKey.icon] as? String,
      let fetchedAtMs = (payload[WatchWeatherKey.fetchedAtMs] as? NSNumber)?.int64Value
    else { return nil }

    let minutes = (payload[WatchWeatherKey.hourMinutes] as? [NSNumber])?.map(\.intValue) ?? []
    let temps = (payload[WatchWeatherKey.hourTemps] as? [NSNumber])?.map(\.intValue) ?? []
    let icons = payload[WatchWeatherKey.hourIcons] as? [String] ?? []
    let precips = (payload[WatchWeatherKey.hourPrecips] as? [NSNumber])?.map(\.intValue) ?? []
    var hourly: [WatchWeatherHour] = []
    for (index, minuteOfDay) in minutes.enumerated() {
      guard index < temps.count, index < icons.count else { continue }
      hourly.append(
        WatchWeatherHour(
          minuteOfDay: minuteOfDay,
          temperatureC: temps[index],
          icon: icons[index],
          precipitationProbability: index < precips.count ? precips[index] : 0
        )
      )
    }

    return WatchWeather(
      temperatureC: temperatureC,
      icon: icon,
      label: payload[WatchWeatherKey.label] as? String ?? "",
      precipitationProbability: (payload[WatchWeatherKey.precipitationProbability] as? NSNumber)?.intValue ?? 0,
      hourly: hourly,
      sunriseMinuteOfDay: (payload[WatchWeatherKey.sunrise] as? NSNumber)?.intValue,
      sunsetMinuteOfDay: (payload[WatchWeatherKey.sunset] as? NSNumber)?.intValue,
      latitude: (payload[WatchWeatherKey.latitude] as? NSNumber)?.doubleValue,
      longitude: (payload[WatchWeatherKey.longitude] as? NSNumber)?.doubleValue,
      fetchedAtMs: fetchedAtMs
    )
  }

  /// The weather channel of a whole Application Context. An absent channel is no forecast, which is
  /// also what a wrist reads before the phone has ever seen a GPS Fix.
  static func decode(context: [String: Any]) -> WatchWeather? {
    decode(context[watchWeatherChannel] as? [String: Any])
  }

  /// Whether this forecast is still worth showing. The Application Context keeps the last value
  /// forever, so a phone that stopped refreshing (app killed, out of range, GPS off) would
  /// otherwise leave yesterday's conditions on the wrist looking current.
  ///
  /// A clock that jumped backwards (timezone or NTP correction) must not read as "from the future"
  /// and hide a forecast that is fine; only real age hides it.
  ///
  /// @parity /watch/wearos/src/main/java/app/vescape/wear/WatchWeather.kt `freshWeather`
  func isFresh(nowMs: Int64) -> Bool { nowMs - fetchedAtMs <= watchWeatherStaleMs }
}

/// How long a pushed forecast is worth showing. Generous next to the phone's ten-minute refresh:
/// this is the line between "a bit old" and "not weather any more".
///
/// @parity /watch/wearos/src/main/java/app/vescape/wear/WatchWeather.kt `WEATHER_STALE_MS`
let watchWeatherStaleMs: Int64 = 3 * 60 * 60 * 1_000

/// Two forecasts that look identical on the wrist are the same push, so the comparison deliberately
/// leaves out `fetchedAtMs` — refetching the same numbers ten minutes later is not a redraw.
///
/// @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/watch/WatchWeather.kt `equals`
extension WatchWeather: Equatable {
  static func == (lhs: WatchWeather, rhs: WatchWeather) -> Bool {
    lhs.temperatureC == rhs.temperatureC
      && lhs.icon == rhs.icon
      && lhs.label == rhs.label
      && lhs.precipitationProbability == rhs.precipitationProbability
      && lhs.hourly == rhs.hourly
      && lhs.sunriseMinuteOfDay == rhs.sunriseMinuteOfDay
      && lhs.sunsetMinuteOfDay == rhs.sunsetMinuteOfDay
      && lhs.latitude == rhs.latitude
      && lhs.longitude == rhs.longitude
  }
}

/// Which sun event the rider still has ahead of them, and when.
///
/// @parity /watch/wearos/src/main/java/app/vescape/wear/WeatherScreen.kt `SunEvent`
struct WatchSunEvent: Equatable {
  var minuteOfDay: Int
  var rising: Bool
}

/// The next sun event from now: sunrise while the sun is still down, sunset once it is up, and
/// tomorrow's sunrise after dark. Nil when the phone sent neither time.
///
/// @parity /watch/wearos/src/main/java/app/vescape/wear/WeatherScreen.kt `nextSunEvent`
func watchNextSunEvent(
  nowMinuteOfDay: Int,
  sunriseMinuteOfDay: Int?,
  sunsetMinuteOfDay: Int?
) -> WatchSunEvent? {
  if let sunriseMinuteOfDay, nowMinuteOfDay < sunriseMinuteOfDay {
    return WatchSunEvent(minuteOfDay: sunriseMinuteOfDay, rising: true)
  }
  if let sunsetMinuteOfDay, nowMinuteOfDay < sunsetMinuteOfDay {
    return WatchSunEvent(minuteOfDay: sunsetMinuteOfDay, rising: false)
  }
  if let sunriseMinuteOfDay { return WatchSunEvent(minuteOfDay: sunriseMinuteOfDay, rising: true) }
  guard let sunsetMinuteOfDay else { return nil }
  return WatchSunEvent(minuteOfDay: sunsetMinuteOfDay, rising: false)
}

/// Minutes since local midnight for an instant, the unit every wrist time is formatted from.
///
/// @parity /watch/wearos/src/main/java/app/vescape/wear/WatchWeather.kt `minuteOfDay`
func watchMinuteOfDay(epochMs: Int64, calendar: Calendar = .current) -> Int {
  let date = Date(timeIntervalSince1970: Double(epochMs) / 1000)
  let parts = calendar.dateComponents([.hour, .minute], from: date)
  return (parts.hour ?? 0) * 60 + (parts.minute ?? 0)
}

/// `HH:MM` for a minute-of-day. Always 24 h, matching Wear OS, so a forecast hour and a frame time
/// never disagree about what "13:00" is.
///
/// @parity /watch/wearos/src/main/java/app/vescape/wear/WatchWeather.kt `formatHour`
func watchFormatHour(_ minuteOfDay: Int) -> String {
  // Floor-mod: a negative minute-of-day can only come from a corrupt bag, and `%` would print a
  // minus sign into a clock face.
  let wrapped = ((minuteOfDay % (24 * 60)) + 24 * 60) % (24 * 60)
  return String(format: "%02d:%02d", wrapped / 60, wrapped % 60)
}
